modules/oidcsso/src/main/java/org/wso2/carbon/tomcat/oidcsso/extension/agent/OIDCAgent.java

```diff
@@ -0,0 +1,367 @@
+            throw new TokenException("Error occurred while obtaining the public key.", e);
+        }
+        Map<String, String> jwkMap = new HashMap<>();
+        jwkMap.put(Constants.KEY_TYPE, publicKey.getAlgorithm());
+        jwkMap.put(Constants.X509_CERTIFICATE_SHA_1_THUMBPRINT, Constants.X509_CERTIFICATE_SHA_1_THUMBPRINT_VALUE);
+        jwkMap.put(Constants.KEY_ID, Constants.KEY_ID_VALUE);
+        jwkMap.put(Constants.ALGORITHM, Constants.ALGORITHM_TYPE);
+        jwkMap.put(Constants.MODULUS, String.valueOf(Base64URL.encode(publicKey.getModulus())));
+        jwkMap.put(Constants.EXPONENT, String.valueOf(Base64URL.encode(publicKey.getPublicExponent())));
+        JWK jwk;
+        try {
+            jwk = JWK.parse(new JSONObject(jwkMap));
+        } catch (java.text.ParseException e) {
+            throw new TokenException("Error occurred while generating the JKW.", e);
+        }
+        JWKSet jwkSet = new JWKSet(jwk);
+        IDTokenValidator idTokenValidator = new IDTokenValidator(issuer, clientID, jwsAlg, jwkSet);
+        IDTokenClaimsSet claimsSet;
+        try {
+            claimsSet = idTokenValidator.validate(oidcTokenResponse.getOIDCTokens().getIDToken(), null);
+        } catch (BadJOSEException | JOSEException e) {
+            throw new TokenException("Error occurred while validating the ID Token.", e);
+        }
+        return claimsSet;
+    }
+
+    /**
+     * this method generates User Info request, makes direct call to User Info endpoint at authorization server,
+     * gets the user info response, processes it and return it.
+     *
+     * @param oidcConfiguration web application specific OIDC configuration
+     * @param acsToken          a string value which is received in token response
+     * @return UserInfoResponse Object
+     * @throws UserInfoException if any error occurs during getting the user information
+     */
+    public UserInfoResponse getUserInfo(OIDCConfiguration oidcConfiguration, String acsToken)
+            throws UserInfoException {
+        BearerAccessToken accessToken = new BearerAccessToken(acsToken);
+        URI userInfoEndpoint = oidcConfiguration.getUserInfoEndpoint();
+        UserInfoRequest userInfoRequest = new UserInfoRequest(userInfoEndpoint, accessToken);
+        HTTPRequest request = userInfoRequest.toHTTPRequest();
+        return processUserInfoResponse(request);
+    }
+
+    /**
+     * this is a default method which sends the user info request and processes the received response which is called by
+     * the method getUserInfo.
+     *
+     * @param request generated user info request
+     * @return UserInfoResponse object
+     * @throws UserInfoException if any errors occur during getting the user info response
+     */
+    UserInfoResponse processUserInfoResponse(HTTPRequest request) throws UserInfoException {
+        UserInfoResponse userInfoResponse = new UserInfoResponse();
+        com.nimbusds.openid.connect.sdk.UserInfoResponse response;
```

Comment:
Do we need this kind of reference here? Any reason not to import this from the top of the class?